core/shared/src/main/scala/coursier/util/Print.scala

```diff
@@ -80,32 +109,36 @@ object Print {
     reverse: Boolean,
     colors: Boolean
   ): String = {
+    val colorsCase = Colors.get(colors)
 
-    val (red, yellow, reset) =
-      if (colors)
-        (Console.RED, Console.YELLOW, Console.RESET)
-      else
-        ("", "", "")
+    if (reverse) {
+      reverseTree(resolution.dependencies.toSeq, resolution, printExclusions).render(_.repr(colorsCase))
+    } else {
+      normalTree(roots, resolution, printExclusions).render(_.repr(colorsCase))
+    }
+
+  }
 
-    final case class Elem(dep: Dependency, excluded: Boolean) {
+  private def getElemFactory(resolution: Resolution, withExclusions: Boolean): Dependency => Elem = {
+    final case class ElemImpl(dep: Dependency, excluded: Boolean) extends Elem {
 
-      lazy val reconciledVersion = resolution.reconciledVersions
+      val reconciledVersion: String = resolution.reconciledVersions
         .getOrElse(dep.module, dep.version)
 
-      lazy val repr =
+      def repr(colors: Colors): String =
```

Comment:
Potentially there could be performance regression changing from `val` to `def` since the value could be access multiple times.

fine with `lazy val` -> `val` tho.

Reply:
Since now the `Tree` and the `Rendering` are two different operations, I had to change it to a method. I'm not sure _how much_ of a performance regression it is, since in the same code path we're doing I/O operations :smile:

Reply:
It is unrelated to IO. `def repr` would be the `show` in https://github.com/coursier/coursier/blob/10f3a067fe93cf209e3dfd72f3bf834bdf73666f/core/shared/src/main/scala/coursier/util/Tree.scala#L39

so each time calling `show(elem)` will run the body of `def repr`. Since `elem` can show up in multiple parts of the tree, it would be doing redundant work.

Reply:
I'm aware it's unrelated to IO, just saying that it runs after lots of IO operations, so the performance regression is negligible...

Reply:
Sure. But reverting to val should be easy then? Usually we trade performance for something, but I fail to see what's gained here.

Reply:
Memory, instead of saving all the strings in the tree, we can output them in a streaming manner.

Reply:
Sure. I'm open to this change when streaming is implemented.

Reply:
@wisechengyi So, what would you like me to do? pass `Colors` along the tree nodes and change `repr` back to `val`? I'm open to changes, I just want to get this pull request merged so I can pass on to implement the `whatDependsOn` operation in mill :smile:

Reply:
Ah okay sorry I missed there's an arg to repr now. Lgtm then. The separation is nice. Thanks

Reply:
:smile: See other comments, Rest is fixed. If you want me to squash all commits I'll do it.

Reply:
Thanks. Should be no need to. We will squash in the end merge anyway.

Reply:
@wisechengyi :+1: 

When's the next release? :smile:

Reply:
that'll be question for @alexarchambault